Modify httpBatchStore so that writing values maintains some locality

cmasone-attic opened the pull request:
NBS benefits from related chunks being near one another. Initially,
let's use write-order as a proxy for "related".

This patch contains a pretty heinous hack to allow sync to continue
putting chunks into httpBatchStore top-down without breaking
server-side validation. Work to fix this is tracked in #2982

This patch fixes #2968, at least for now

cmasone-attic committed:
Modify httpBatchStore so that writing values maintains some locality

NBS benefits from related chunks being near one another. Initially,
let's use write-order as a proxy for "related".

This patch contains a pretty heinous hack to allow sync to continue
putting chunks into httpBatchStore top-down without breaking
server-side validation. Work to fix this is tracked in #2982

This patch fixes #2968, at least for now

cmasone-attic commented:
PTAL

Now that I've gone all the way through the patch, I'm not sure that having NomsBlockCache be distinct from NomsBlockStore makes a lot of sense. On one hand, it's a nice place to put the code that creates and deletes the temp directory that the cache data is in, and it means that `ExtractChunks` is not part of the NomsBlockStore public API. On the other hand...it's a pretty thin wrapper, and the guts of `extractChunks` really needs to be inside NomsBlockStore so it can get at `nbs.tables` and the mutex that protects it.

I'm trying a little bit of further cleanup, but I broke something while doing so. I'll continue trying to chase that down, but please go ahead and review

rafael-atticlabs commented on go/datas/http_batch_store.go:
```diff
@@ -313,16 +332,21 @@ func resBodyReader(res *http.Response) (reader io.ReadCloser) {
 
 	bhcs.requestWg.Add(1)
-	bhcs.writeQueue <- writeRequest{c.Hash(), hints, false}
+	bhcs.writeQueue <- writeRequest{hints, false}
```
I'm wondering whether the machinery around the write queue (the channel, the persistent goroutine that are servicing the write-queue and the flush queue) are just unneeded complexity now. this stuff is a hold over from the before time when we didn't have any validation and thus were writing in larger batches, but we haven't done that in a while.

i think it's possible now to just have `SchedulePut` Put into the nbs (not care about whether the chunk was contained), and add its hints to the hint set, all inside the mutex, and have `Flush` just send everything over.

i'm probably forgetting about some critical thing, right?

cmasone-attic replied:
No, I think this will work. As long as  `AddHints()` uses that mutex also, I think it will work.

rafael-atticlabs commented on go/datas/http_batch_store.go:
```diff
@@ -448,8 +470,7 @@ func (bhcs *httpBatchStore) Root() hash.Hash {
 	// POST http://<host>/root?current=<ref>&last=<ref>. Response will be 200 on success, 409 if current is outdated.
-	bhcs.Flush()
-
+	bhcs.flushInternal(nbs.InsertOrder)
```
So is it the case that the sync user always sends over chunks by calling Flush and the simple write user always calls UpdateRoot()?

If so, I guess I'm ok with this, but it is a bit mysterious. If so, let's comment it somewhere and maybe put more detail in the bug you opened that the specific problem is that the API of http_batch_store is now kind of wonky.

As an alternative, maybe BatchStore gets a temporary method called "MarkFlushReverseOrder", or something that only sync calls before it starts writing top-down?

cmasone-attic replied:
This is the thing I was working on changing after I sent this patch, because I also don't like that this relies on making `noms sync` only thing that calls BatchStore.Flush(). This patch made that so, but there's no guarantee at all that it would remain that way.

Something mysteriously broke when I changed things that way, so I'll need to chase that down

cmasone-attic commented on go/nbs/cache.go:
```diff
@@ -0,0 +1,62 @@
+	dbDir  string
+}
+
+// Insert stores c in the cache. If c is successfully added to the cache,
```
you're right. Maybe I don't actually need these semantics if I can get rid of all the goroutine machinery

cmasone-attic commented on go/nbs/mem_table.go:
```diff
@@ -79,6 +79,20 @@ func (mt *memTable) getMany(reqs []getRecord, wg *sync.WaitGroup) (remaining boo
 }
 
+func (mt *memTable) extract(order EnumerationOrder, chunks chan<- extractRecord) {
+	sort.Sort(hasRecordByOrder(mt.order)) // ensure "insertion" order for extraction
```
I think it will be in insertion order, but we don't document that as a requirement, so I was programming defensively. I can just require that that be an invariant

cmasone-attic commented on go/nbs/table_reader.go:
```diff
@@ -424,3 +424,34 @@ func (tr tableReader) calcReads(reqs []getRecord, blockSize, maxReadSize, ampThr
+	d.Chk.True(uint64(n) == chunkLen)
+
+	sendChunk := func(i uint32) {
+		localOffset := tr.offsets[i] - tr.offsets[0]
```
Opened https://github.com/attic-labs/noms/issues/2986

cmasone-attic commented on go/nbs/table_set.go:
```diff
@@ -96,6 +96,19 @@ func (ts tableSet) Prepend(mt *memTable) tableSet {
+		}
+		return
+	}
+	for i := len(ts.chunkSources) - 1; i >= 0; i-- {
```
so would I!

Go, however, apparently doesn't. Increment and decrement aren't expressions in Go (statements? whatever the right term is. You can't use them inline like that :-/)

rafael-atticlabs replied:
:-)

cmasone-attic committed:
Introduces PullWithFlush() to allow noms sync to explicitly
pull chunks over and flush directly after. This allows UpdateRoot
to behave as before.

Also clears out all the legacy batch-put machinery. Now, Flush()
just directly calls sendWriteRequests().

cmasone-attic commented:
Actually, can you look again? I've squashed so that the second commit includes all the changes since yesterday, which includes getting rid of the old batch-put machinery, that `flushInternal()` hack, and some other things.

rafael-atticlabs commented:
Still LG
